doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1748,31 +1757,39 @@ Controlling the proof flow
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction.  If
+   :n:`@one_term_with_bindings` is not provided, the tactic first does a
+   :tacn:`intros`
+   then succeeds if there is a hypothesis :n:`False` in the updated context, or
+   if there is a pair of hypotheses where one is simple negation of the other
+   (i.e. :n:`P` and not :n:`~P`).
 
-   .. exn:: No such assumption.
-      :undocumented:
+   Use the :tacn:`discriminate` tactic to prove the current goal when there
+   is a hypothesis with an impossible structural equality such as
+   :n:`0 = 1`.
 
-.. tacv:: contradiction @ident
+   If :n:`@one_term_with_bindings` has the type :n:`False`, the tactic
+   proves the current goal.  If the type is a negation, e.g. :n:`~P`, then
+   the current goal is replaced with :n:`P`.  (However, the converse doesn't
+   work, i.e. if the type is :n:`P` and the goal is :n:`~P`.)
 
-   The proof of False is searched in the hypothesis named :n:`@ident`.
+   .. todo PR: examples?  Or maybe it's simple enough that they're not needed
```

Comment:
Sure, examples are always a good thing to provide. However, I would stay mostly with very simple examples (`contradiction` with no arguments) given how limited the usefulness of the variant with arguments is. The only example of `contradiction` with arguments that might be worth to provide is the following (adapted from the stdlib):

```coq
Require Import Arith.

Goal forall (A : Prop), 0 < 0 -> A.
intros. (* In this case, the call to intros is necessary *)
contradiction (Nat.lt_irrefl 0).
Qed.
```

Some simple examples with no argument:

```coq
Inductive F :=. (* Another empty inductive type *)

Goal F -> False.
contradiction.
Qed.
```

```coq
Goal forall (A : Prop), A -> ~A -> False.
contradiction.
Qed.
```

```coq
Goal forall (A : Type) (x : A), ~(x = x) -> False.
contradiction.
Qed.
```

Reply:
The `intros` is necessary in the first example (`Goal forall (A : Prop), 0 < 0 -> A.`) because the argument to the tactic must be `~P` to match `P`.  Seems to me that tactic would be much simpler for users if the argument could also be `P` to match `~P`.  I can file an issue.  WDYT?